### PR TITLE
Stop ignoring some peers when updating the address book

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Zebra was ignoring some peers when updating the address book.

This could be the cause of issues like #3042.

## Solution

- Correctly apply MetaAddrChanges when there is no existing address book entry
- Add all initial peers to the address book (not just the unused ones)

## Review

This PR is ready for review.

I think @dconnolly can review this PR.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

I've tested this PR locally, and the crawler and address book metrics look a lot better. The number of peer addresses rises rapidly, and gets full after a few hours. And Zebra doesn't hang any more, even after a few days.

### CODE REVIEW

**Comment by teor2345** [comment]
I've tested this PR locally, and the crawler and address book metrics look a lot better. The number of peer addresses rises rapidly, and gets full after a few hours. And Zebra doesn't hang any more, even after a few days.

**Comment by teor2345** [comment]
Marking as a high priority, because this bug sometimes makes CI fail.

File: zebra-network/src/peer_set/initialize.rs [hunk 1]
```diff
@@ -375,37 +375,40 @@ where
 /// Limit the number of `initial_peers` addresses entries to the configured
 /// `peerset_initial_target_size`.
 ///
-/// Returns randomly chosen entries from the provided set of addresses.
+/// Returns randomly chosen entries from the provided set of addresses,
+/// in a random order.
 ///
-/// Sends any unused entries to the `address_book_updater`.
+/// Also sends every initial peer to the `address_book_updater`.
```
Comment (dconnolly) [inline zebra-network/src/peer_set/initialize.rs:381]: 👍

File: zebra-network/src/peer_set/initialize.rs [hunk 2]
```diff
@@ -375,37 +375,40 @@ where
...
     let all_peers = config.initial_peers().await;
     let peers_count = all_peers.len();
 
-    if peers_count <= config.peerset_initial_target_size {
-        return all_peers;
-    }
+    // # Correctness
+    //
+    // We can't exit early if we only have a few peers,
+    // because we still need to shuffle the connection order.
```
Comment (dconnolly) [inline zebra-network/src/peer_set/initialize.rs:392]: 👍

File: zebra-network/src/peer_set/initialize.rs [hunk 3]
```diff
@@ -375,37 +375,40 @@ where
...
+            peers_count, config.peerset_initial_target_size
+        );
+    }
 
-    // Split all the peers into the `initial_peers` that will be returned and
-    // `unused_peers` that will be sent to the address book.
-    let mut all_peers: Vec<SocketAddr> = all_peers.into_iter().collect();
-    let (initial_peers, unused_peers) =
-        all_peers.partial_shuffle(&mut rand::thread_rng(), config.peerset_initial_target_size);
+    // Split out the `initial_peers` that will be shuffled and returned.
```
Comment (dconnolly) [inline zebra-network/src/peer_set/initialize.rs:401]: 👍

File: zebra-network/src/peer_set/initialize.rs [hunk 4]
```diff
@@ -375,37 +375,40 @@ where
...
+    let mut initial_peers: Vec<SocketAddr> = all_peers.iter().cloned().collect();
+    let (initial_peers, _unused_peers) =
+        initial_peers.partial_shuffle(&mut rand::thread_rng(), config.peerset_initial_target_size);
 
-    // Send the unused peers to the address book.
-    for peer in unused_peers {
-        let peer_addr = MetaAddr::new_initial_peer(*peer);
+    // Send every initial peer to the address book.
+    // (This treats initial peers the same way we treat gossiped peers.)
```
Comment (dconnolly) [inline zebra-network/src/peer_set/initialize.rs:407]: 👍

File: zebra-network/src/meta_addr/tests/prop.rs [hunk 5]
```diff
@@ -174,10 +174,107 @@ proptest! {
             canonical_local_listener,
         );
     }
+
+    /// Make sure that [`MetaAddrChange`]s are correctly applied
+    /// when there is no [`MetaAddr`] in the address book.
```
Comment (dconnolly) [inline zebra-network/src/meta_addr/tests/prop.rs:179]: 👍

**Review by dconnolly — APPROVED** [review]
lgtm!

**Pull request merged**